design/authorization-handler.md

```diff
@@ -0,0 +1,590 @@
+
+For example [Grant Management API specification - section 6.4](https://openid.net/specs/fapi-grant-management-01.html#section-6.4) (not yet supported by Keycloak) allows clients to query Keycloak about the granted consents. Note that the returned structure contains both scope and authorization_details and hence it is needed by KC to track how the particular request was sent.
+
+### AuthorizationContext object
+
+Output of the parsing is AuthorizationContext object, which is a list of AuthorizationDetailsEntry objects. Each AuthorizationDetailsEntry contains data like:
+
+* JSON with the authorization details
+* AuthorizationHandler used
+* Source - the parsing method (scope parameter or authorization_details parameter)
+
+#### Java pseudo-structure
+
+```java
+class AuthorizationContext {
+	/**
+ 	*  List of entries of parsed values from "scope" parameter and    
+      * from "authorization_details" .
+ 	*  Each entry represents single parsed value from "scope"
+      * parameter or single entry of "authorization_details"
+ 	*
+ 	* There are also entries corresponding to all "default client
+     * scopes" for the particular client, which are added
+     * automatically.
+ 	*/
+	private List<AuthorizationDetailsEntry> entries;
+}
+
+class AuthorizationDetailsEntry {
+    	// Client scope linked with this authorizationDetails
+    	ClientScopeModel clientScope;
+
+    	// How this entry was added. See enum "Source" below
+    	Source source;
+
+    	// AuthorizationDetails data sent by the client. In case that
+// this was added by "scope" parameter, the "scope" entry is wrapped
+// to the authorizationDetails. See below for the details:
+      AuthorizationDetailsJSONRepresentation authzDetails;  	 
+
+}
+
+enum Source {
+    	// Entry added with "scope" parameter
+    	SCOPE_PARAMETER,
+
+    	// Entry added with "authorization_details" parameter
+    	AUTHORIZATION_DETAILS_PARAMETER,
+
+    	// Entry added automatically by default client scope
+    	DEFAULT_SCOPE   	 
+}
+ 	 
+class AuthorizationDetailsJSONRepresentation {
+   	// These fields are from RAR specification
+   	private String type;
+   	private List<String> locations;
+   	private List<String> actions;
+   	private List<String> datatypes;
+   	private String identifier;
+   	private List<String> privileges;
+  	 
+   	// Custom authorizationDetails data for particular type
+   	private Map<String, Object> otherData;
+}
+```
+
+
+## Parsing
+
+Assumption of this approach is that there is no limitation on Authorization Handler whether it can be added to the request by “scope” parameter or by “authorization_details” parameter.
+
+Requests will go through the chain of parsers. By default, we will have three parsers defined:
```

Comment:
It might be good if you would write how to distinguish static scope and dynamic scope during the chain of parsers.

Reply:
@tnorimat @dgozalo I prefer to the alternative. The logic of how to distinguish static scope and dynamic scope can be 
 enclosed onto it, not needed to be treated separately with parsing.

Reply:
I agree with @tnorimat on this. The most we can avoid touching the current data in the database, the better. If new, dynamic scopes will be created with a new, regexp field, it should not impact the already existing scopes at all. Otherwise they would need to be migrated to have a valid regexp in the database.